Fix bugs: ВР-127, ВР-128, BP-131 and /administrators/refresh endpoint

ivr42 opened the pull request:
- administrator role and status change endpoints now get values of role or status from the request body
  - request models `AdministratorRoleRequest` and `AdministratorStatusRequest` was added according to validate the status and role from the request body
  - method `check_administrator_existence` of the `AdministratorRepository` class was replaced by the `administrator_exists` with the `role` argument;
  - method `check_administrator_by_token` was added to the `AuthenticationService` class;
  - method `check_administrator_by_token` is used now instead of the `get_current_active_administrator` in the places where you need to check only;
  - in method `restore_administrator_password` of class `AdministratorService` additional check was added: you are not allowed to change the password if you don't have the `Administrator.Role.ADMINISTRATOR` role;
  - in methods `deactivate_invitation` and `reactivate_invitation` of class `AdministratorInvitationCBV` additional check was added: you are not allowed to do these actions if you don't have the `Administrator.Role.ADMINISTRATOR` role;
  - in method `refresh` of the `AdministratorCBV` class authorization was removed;
  - unnecessary exceptions were removed.  

Задачи в Notion: 
- [ВР-127](https://www.notion.so/127-0e9ac4f376254044abe278d2d5f1ace4)
- [ВР-128](https://www.notion.so/128-d3f6c10dabe84edf908d97704d38645e)
- [BP-131](https://www.notion.so/131-373ff7be88994e8d9b35da2e073696b4)
- [/administrators/refresh endpoint](https://www.notion.so/Fix-GET-administrators-refresh-660824ba402f4b29a5a0779789d52dab)

ivr42 committed:
Bub BP-131 has been fixed.

  - in method `restore_administrator_password` of class
    `AdministratorService` additional check was added: you are not allowed
    to change the password if you don't have the role of
    `Administrator.Role.ADMINISTRATOR` or if you are not changing the
    password yourself.
  - Exception `AdministratorRestPasswordError` was added

ADesBiysk reviewed (commented):
Здесь нужно будет тоже ориентироваться на решение в данном ПР https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/319#pullrequestreview-1402388010

ivr42 committed:
Refactoring in AdministratorService

  - method `get_email_from_token` was added to the
    `AuthenticationService` class
  - method `switch_a_field_with_checks` was added to the
    `AdministratorService` class instead of methods
    `block_administrator` and `switch_administrator_role`
  - method `check_administrator_is_admin_by_email` was added to the
    `AdministratorService` class
  - `AuthenticationService` and `AdministratorService` are coupled now
  - method `administrator_is_active_and_is_admin` added to the
    `AdministratorRepository` class
  - methods which are swith administrator status and role are moved to
    the `Administrator` model from `AdministratorService` class

ivr42 committed:
administrator endpoint has been refactored

ivr42 committed:
Fix: endpoint `/administrators/refresh` available without access_token now

ivr42 committed:
Refactoring:

  - method `administrator_exists` of class `AdministratorRepository`
    checks all conditions and used instead of methods:
    `administrator_is_active` and `administrator_is_active_and_is_admin`
  - method `check_administrator_by_token` was added to the class
    `AuthenticationService`.
  - `check_administrator_by_token` is used in administrator and
    administrator_invitation routers to check administrator permissions.

AntonRev commented on src/core/services/authentication_service.py:
```diff
@@ -77,19 +77,44 @@ async def login(self, auth_data: AdministratorAuthenticateRequest) -> Administra
+            token (str): JWT token.
+            is_active (bool): True — проверяет, что администратор активный;
+                              False — проверяет, что администратор не активный;
+                              None — статус администратора не проверяется.
```
* Нам нигде не нужны не активные пользователи, этот флаг лучше убрать.
* Надо еще посмотреть другие роуты и там поправить проверку администратора

ivr42 replied:
OK. Сегодня сделаю

ivr42 committed:
Code review admonitions were fixed.

  - method `check_administrator_by_token` is used now instead of the
    get_current_active_administrator in the places where you need to
    check only;
  - unnecessary argument `is_active` was removed from the
    `check_administrator_by_token` method of the `AuthenticationService`
    class;
  - unnecessary argument `status` was removed from the
    `administrator_exists` method of the `AdministratorRepository`
    class;
  - unused exception were removed.

ivr42 committed:
Merge branch 'develop' into fix/bug_BP-131

ADesBiysk commented on src/core/db/repository/administrator_repository.py:
```diff
@@ -26,15 +26,31 @@ async def get_by_email(self, email: str) -> Administrator:
 
-    async def check_administrator_existence(self, email: str) -> bool:
-        """Проверяет существование администратора по email.
+    async def administrator_exists(
```
is_administrator_exists?

ivr42 replied:
ок

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -246,17 +248,15 @@ class InvitationAlreadyActivatedError(BadRequestError):
-class AdministratorChangeError(ForbiddenError):
-    detail = "У вас нет прав на изменение других администраторов."
+class AdministratorSelfChangeError(ForbiddenError):
+    detail = "Вы не можете изменить самого себя."
```
Какое-то философское и депрессивное сообщение об ошибке. :)

ivr42 replied:
Да, вообще поменял ошибки, эту удалил, а сделал две отдельных на изменение статуса и изменение роли

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -246,17 +248,15 @@ class InvitationAlreadyActivatedError(BadRequestError):
-    detail = "Вы не можете заблокировать себя."
+class AdministratorUnknownRoleError(BadRequestError):
+    def __init__(self, role: Administrator.Role):
+        self.detail = "Неизвестный статус администратора: {}".format(role)
```
Роль

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -189,36 +184,36 @@ async def administrator_change_role(
-        """Заблокировать администратора."""
-        current_admin = await self.authentication_service.get_current_active_administrator(token.credentials)
-        return await self.administrator_service.block_administrator(current_admin, administrator_id)
+        """Изменить статус администратора."""
```
С фронтом обсудили и решили разделить эндпойнты.

PATCH /administrators/{administrator_id}/block
PATCH /administrators/{administrator_id}/unblock

PATCH /administrators/{administrator_id}/role/expert
PATCH /administrators/{administrator_id}/role/administrator

ivr42 replied:
Как предложение, может сделать так:

PATCH /administrators/{administrator_id}/status/active
PATCH /administrators/{administrator_id}/status/blocked

1. будет одинаково с изменением роли
2. мы уйдём от глаголов в эндпоинтах, что положительно (у нас же уже есть глагол — PATCH)

Или может вообще передавать их не как path parameters, а в теле (body)?
_Это и для статуса, и для роли_

ADesBiysk replied:
Решили сделать 2 эндпойнта: status и role, а значения передавать через body.

ivr42 replied:
ok

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -65,23 +77,51 @@ async def login(self, auth_data: AdministratorAuthenticateRequest) -> Administra
+    async def check_administrator_by_token(
+        self,
+        token: HTTPAuthorizationCredentials,
+        is_admin: bool | None = None,
```
Всё-таки лучше сюда сразу роль передавать. Если потребуется добавить третью роль, то эту функцию и все её вызовы нужно будет переделывать, что не очень хорошо.

ivr42 replied:
ок

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -65,23 +77,51 @@ async def login(self, auth_data: AdministratorAuthenticateRequest) -> Administra
+
+        check_conditions = {"email": email}
+
+        if is_admin:
```
Мне кажется, или эти 2 условия, в данном случае, идентичны? :)
```
if is_admin:
if is_admin is not None:
```

ivr42 replied:
не, в моём случае оно разное
может быть `True` — тогда работает ветка `if is_admin `, `False` — ветка `if is_admin is not None ` и `None` — тогда ничего не добавляется. 

Но в случае, если мы передаём туда прямо роль `Administrator.Role`, будет проще — как ты и сказал эти два условия будут идентичны

ivr42 committed:
Code review fixes:

  - administrator role change router was changed from switcher to setter
  - administrator status change router was changed from switcher to
    setter
  - method `check_administrator_by_token` now get an
    `Administrator.Role` argument
  - method `administrator_exists` renamed to `is_administrator_exists`

ADesBiysk commented on src/core/db/repository/administrator_repository.py:
```diff
@@ -26,15 +26,31 @@ async def get_by_email(self, email: str) -> Administrator:
-        administrator_exists = await self._session.execute(
-            select(select(Administrator).where(Administrator.email == email).exists())
-        )
+        check_conditions = [
```
Предлагаю сразу формировать запрос постепенно уточняя его. Такой подход мне кажется более, очевидным и наглядным.

ADesBiysk commented on src/core/services/administrator_service.py:
```diff
@@ -95,18 +76,38 @@ async def __set_new_password(self, password: str, email: str) -> Administrator:
+    async def change_administrator_status(
+        self, administrator_id: UUID, status: Administrator.Status, changer_token: str
+    ) -> Administrator:
+        if status not in Administrator.Status.__members__.values():
```
Не уверен, что такая проверка нужна. Мы указали тип аргумента в аннотации, и если программист намерено передаёт что-то другое, то функция должна упасть. То, что это будет наше исключение, а не системное, мало что изменит.
Если переделать на передачу значения через тело запроса, то там валидацией будет заниматься pydantic.

ivr42 replied:
ok

ivr42 committed:
Code review fixes:

  - administrator role change endpoint now gets role from the request
    body
  - administrator status change endpoint now gets status from the
    request body
  - request models `AdministratorRoleRequest` and
    `AdministratorStatusRequest` was added according to validate status
    and role from the request body
  - some refactoring has been done

ivr42 committed:
Fix summary phrase in administrator status endpoint
